addons/mail/tests/test_mail_channel_as_guest.py

```diff
@@ -4,13 +4,38 @@
 import odoo
 from odoo.tests import HttpCase
 
+
 @odoo.tests.tagged('-at_install', 'post_install')
-class TestMailGuestPages(HttpCase):
-    def test_mail_channel_as_guest(self):
-        """Checks that the invite page redirects to the channel and that all
-        modules load correctly on the welcome and channel page"""
-        channel = self.env['mail.channel'].create({
+class TestMailPublicPage(HttpCase):
+    """Checks that the invite page redirects to the channel and that all
+    modules load correctly on the welcome and channel page when authenticated as various users"""
+
+    def setUp(self):
+        super().setUp()
+        self.channel = self.env['mail.channel'].create({
             'name': 'Test channel',
             'public': 'public',
         })
-        self.start_tour(f"/chat/{channel.id}/{channel.uuid}", "mail/static/tests/tours/mail_channel_as_guest_tour.js")
+        self.route = f"/chat/{self.channel.id}/{self.channel.uuid}"
+        self.tour = "mail/static/tests/tours/discuss_public_tour.js"
+
+    def _open_channel_page_as_user(self, login):
+        self.start_tour(self.route, self.tour, login=login)
+        # Second run of the tour as the first call has side effects, like creating user settings or adding members to
+        # the channel, so we need to run it again to test different parts of the code.
+        self.start_tour(self.route, self.tour, login=login)
+
+    def test_mail_channel_public_page_as_admin(self):
+        self._open_channel_page_as_user('admin')
+
+    def test_mail_channel_public_page_as_guest(self):
+        self.start_tour(self.route, "mail/static/tests/tours/mail_channel_as_guest_tour.js")
+        guest = self.env['mail.guest'].create({'name': 'Guest'})
+        self.channel.add_members(guest_ids=[guest.id])
```

Comment:
guest should be created from the first call of the tour, and should be found by last id, rather than created here